File: PSP_DAM_ACTEVA03B Java/src/restaurante/ListaPedidos.java
package restaurante;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.List;

/**
 * Clase que gestiona la lista de pedidos compartida por los cocineros
 * Contiene la sincronización
 * También se encarga de escribir un log en archivo de texto
 */
public class ListaPedidos {
    private List<Pedido> pedidos;  // Lista compartida
    private int pedidosProcesados; // Contador
    private final File logFile;    // Archivo para log

    public ListaPedidos(List<Pedido> pedidos) {
        this.pedidos = pedidos;
        this.pedidosProcesados = 0;

        // Crear la carpeta "logs" si no existe
        File carpetaLogs = new File("logs");
        if (!carpetaLogs.exists()) {
            carpetaLogs.mkdir();
        }
        // Archivo de log dentro de la carpeta logs
        this.logFile = new File(carpetaLogs, "log_pedidos.txt");
    }

    /**
     * Método sincronizado para que solo un hilo a la vez tome un pedido
     * elimina y retorna el primer pedido de la lista
     * @return Pedido o null si la lista está vacía
     */
    public synchronized Pedido tomarPedido() {
        if (pedidos.isEmpty()) {
            return null;
        }
        return pedidos.remove(0);
    }

    /**
     * Método sincronizado que registra la preparación de un pedido,
     * imprime mensaje en consola y escribe en el archivo de log
     * @param pedido Pedido que se ha preparado
     * @param idCocinero Identificador del hilo/cocinero
     */
    public synchronized void registrarPedido(Pedido pedido, int idCocinero) {
        pedidosProcesados++;
        String mensaje = "Pedido " + pedido.getId() + " (" + pedido.getNombrePlato() +
                         ") preparado por cocinero " + idCocinero;
        System.out.println(mensaje);
        escribirEnLog(mensaje);
    }

    /**
     * Escribe una línea en el archivo de log, con manejo básico de errores
     */
    private void escribirEnLog(String mensaje) {
        try (FileWriter fw = new FileWriter(logFile, true);
             PrintWriter pw = new PrintWriter(fw)) {
            pw.println(mensaje);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: PSP-DAM-ACTEVA03B Python/lista_pedidos.py
import threading

# Clase ListaPedidos: gestiona la lista compartida y el log
class ListaPedidos:
    def __init__(self, pedidos):
        self.pedidos = pedidos
        self.lock = threading.Lock()
        self.log_file = "logs/log_pedidos.txt"
        self._crear_carpeta_logs()

    def _crear_carpeta_logs(self):
        import os
        if not os.path.exists("logs"):
            os.mkdir("logs")

    def tomar_pedido(self):
        with self.lock:
            if not self.pedidos:
                return None
            return self.pedidos.pop(0)

    def registrar_pedido(self, pedido, id_cocinero):
        mensaje = f"Pedido {pedido.id} ({pedido.nombre_plato}) preparado por cocinero {id_cocinero}"
        print(mensaje)
        with self.lock:
            with open(self.log_file, "a", encoding="utf-8") as f:
                f.write(mensaje + "\n")
